File: javascript/cote/GJM/위상정렬.mjs
const TEST_CASE = 10
const MAX_NODE = 100
const [MIN_PER_RANK, MAX_PER_RANK] = [1, 10]
const [MIN_RANKS, MAX_RANKS] = [3, 10]
const PERCENT = 0.7


function initialize() {
	var ret = []
	const rank = Math.floor(Math.random() * (MAX_RANKS - MIN_RANKS)) + MIN_RANKS

	for (let r = 0; r < rank; r++) {
		const new_nodes = Math.floor(Math.random() * (MAX_PER_RANK - MIN_PER_RANK)) + MIN_PER_RANK

		if (ret.length + new_nodes >= MAX_NODE) break

		for (let i = 0; i < ret.length; i++) {
			for (let j = i + 1; j < ret.length + new_nodes; j++) {
				if (Math.random() < PERCENT) continue;
				ret[i].push(j)
			}
		}

		ret = ret.concat(new Array(new_nodes).fill(0).map(i => new Array()))
	}

	return ret;
}

function solve(graph) {

	const visited = new Array(graph.length).fill(false)
	const ret = []

	function dfs(cur) {
		visited[cur] = true

		for (let i of graph[cur]) {
			if (visited[i]) continue
			dfs(i)
		}

		ret.push(cur)
	}


	for (let i = 0; i < graph.length; i++) {
		if (visited[i]) continue
		dfs(i)
	}

	return ret.reverse()
}

for (let it = 0; it < TEST_CASE; it++) {
	const graph = initialize()
	console.log(solve(graph))
}

File: c++/cote/GJM/위상_정렬.cpp
#include <iostream>
#include <memory>
#include <random>
#include <vector>
#define MIN_PER_RANK 1
#define MAX_PER_RANK 10
#define MIN_RANKS 3
#define MAX_RANKS 10
#define TEST_CASES 10
#define MAX_NODE 100
#define PERCENT 30

using namespace std;

vector<vector<int>> initialize(default_random_engine& gen) {
  vector<vector<int>> ret{};

  int ranks = gen() % (MAX_RANKS - MIN_RANKS) + MIN_RANKS;
  int nodes = 0;

  for (int k = 0; k < ranks; k++) {
    int new_nodes = gen() % (MAX_PER_RANK - MIN_PER_RANK) + MIN_PER_RANK;

    if (nodes + new_nodes >= MAX_NODE) {
      break;
    }

    for (int i = 0; i < nodes; i++) {
      for (int j = nodes; j < nodes + new_nodes; j++) {
        if (gen() % 100 <= PERCENT) {
          ret[i].push_back(j);
        }
      }
    }

    for (int i = 0; i < new_nodes; i++) {
      ret.push_back(vector<int>{});
    }
    nodes += new_nodes;
  }

  return ret;
}

void dfs(int cur, vector<vector<int>>& graph, bool* visited,
         vector<int>& path) {
  visited[cur] = true;

  for (int i : graph[cur]) {
    if (!visited[i]) {
      dfs(i, graph, visited, path);
    }
  }

  path.push_back(cur);
}

vector<int> solve(vector<vector<int>>& graph) {
  vector<int> path{};
  bool* visited = new bool[graph.size()];

  for (int i = 0; i < graph.size(); i++) {
    if (!visited[i]) {
      dfs(i, graph, visited, path);
    }
  }

  delete[] visited;

  return path;
}

int main(void) {
  default_random_engine gen{43};

  for (int it = 0; it < TEST_CASES; it++) {
    vector<vector<int>> graph = initialize(gen);
    vector<int> path = solve(graph);

    for (int v : path) {
      cout << v << " ";
    }
    cout << endl;
  }

  return 0;
}